armi/physics/neutronics/fissionProductModel/fissionProductModel.py

```diff
@@ -205,60 +198,146 @@ def _getAllFissionProductNames(self):
                 if fpName not in fissionProductNames:
                     fissionProductNames.append(fpName)
 
-        self.fissionProductNames = fissionProductNames
-
-    def _cacheLFPDensities(self, blockList):
-        # pass 2: Cache all LFP densities for all blocks (so they aren't read
-        # for each FP)
-        runLog.debug("  Caching LFP densities of all blocks")
-        lfpDensities = {}
-        for b in blockList:
-            for lfpName in self._globalLFPs:
-                lfpDensities[lfpName, b.getName()] = b.getNumberDensity(lfpName)
-        return lfpDensities
+        return fissionProductNames
 
-    def updateFissionGasRemovalFractions(self):
+    def removeFissionGasesFromBlocks(self, gasRemovalFractions: dict):
         """
-        Synchronize fission gas removal fractions of all LFP objects with the reactor state.
+        Removes the fission gases from each of the blocks in the core.
 
-        The block parameter ``fgRemoval`` is adjusted by fuel performance
-        modules and is applied here.
-
-        If the ``makeAllBlockLFPsIndependent`` setting is not activated
-        (default), the global lump gets the flux-weighted average of all blocks.
-        Otherwise, each block gets its own fission gas release fraction applied
-        to its individual LFPs. For MCNP restart cases, it is recommended to
-        activate the ``makeAllBlockLFPsIndependent`` setting.
+        Parameters
+        ----------
+        gasRemovalFractions : dict
+            Dictionary with block objects as the keys and the fraction
+            of gaseous fission products to remove.
 
         Notes
         -----
-        The CrossSectionGroupManager does this for each XSG individually for
-        lattice physics, but it's important to keep these up to date as well for
-        anything else that may be interested in fission product information
-        (e.g. MCNP).
-
-        See Also
-        --------
-        armi.physics.neutronics.crossSectionGroupManager.AverageBlockCollection.getRepresentativeBlock
-
+        The current implementation will update the number density vector
+        of each of the blocks and the gaseous fission products are not
+        moved or displaced to another area in the core.
```

Comment:
This is a new implementation of this method that actually changes the number densities in the blocks/components where gaseous fission products should be removed. This call is paired with changes in the fuel performance plugin so that when the fuel performance plugin calculates how much gas to remove in a time-step/node it will only remove the specified amount of gas.

One limitation of this method is that it removes the gaseous fission products from existence rather than migrating them somewhere else. I think that this can be improved later on, but I didn't want to add a bunch of design-specific code that would not be generic here. For example, in some designs the fission gas may migrate to a plenum, but not all designs have this.

I can further clarify this caveat/note in the docstring if that is requested.

Reply:
This probably needs a few eyes on it because there is different treatment that is applied when the lumped fission product model is used versus the `explicitFissionProducts` model. The `explicitFissionsProducts` model is more straight-forward because we are just modifying the gaseous nuclides on the block/component directly. In the lumped fission product model, we instead have `LFP35`, `LFP38`, etc. in the block/component and in this case we need to reach into the lumped fission product class and look at the yields of each gaseous nuclide and then calculate the number density of the gaseous nuclides to subtract from the `LFP35`, `LFP38`, etc. number densities.

I have added unit tests for both of these model types and varied the gas removal fractions from 0.0, 0.5, and 1.0 to check that none, half, and all the gaseous fission products are removed, respectively.